status/service/status.go

```diff
@@ -46,14 +51,94 @@ func (st *Stat) Request(req *Request) (network.Message, error) {
 	}, nil
 }
 
+var errTimeout = errors.New("timeout while waiting for replies")
+
+// CheckConnectivity does an all-by-all connectivity test
+func (st *Stat) CheckConnectivity(req *CheckConnectivity) (*CheckConnectivityReply, error) {
+	// Check signature
+	hash, err := req.hash()
+	if err != nil {
+		return nil, errors.New("couldn't hash message: " + err.Error())
+	}
+	err = schnorr.Verify(cothority.Suite, st.ServerIdentity().Public, hash, req.Signature)
+	if err != nil {
+		return nil, errors.New("signature verification failed: " + err.Error())
+	}
+	if math.Abs(time.Now().Sub(time.Unix(req.Time, 0)).Seconds()) > 120 {
+		return nil, errors.New("too old request")
+	}
+
+	list := req.List
+	id, _ := onet.NewRoster(list).Search(st.ServerIdentity().ID)
+	if id < 0 {
+		return nil, errors.New("cannot check nodes without being in the roster")
+	}
+	if id > 0 {
+		log.Lvl1("Re-arranging list")
+		list[0], list[id] = list[id], list[0]
+	}
+
+	// Test the whole list
+	log.Lvl1("Checking whole roster")
+	to := time.Duration(req.Timeout)
+	err = st.testNodes(list, to)
+	if err == nil {
+		return &CheckConnectivityReply{Nodes: list}, nil
+	}
+
+	if !req.FindFaulty {
+		return nil, errors.New("one or more of the nodes did not reply. " +
+			"Run with FindFaulty=true")
+	}
+
+	// Add one node after the other and only keep nodes that have a full
+	// connectivity
+	newList := []*network.ServerIdentity{list[0]}
+	for _, si := range list[1:] {
+		tmpList := append(newList, si)
+		log.Lvl1("Checking list", tmpList)
+		err = st.testNodes(tmpList, to)
+		if err != nil {
+			log.Warn("Couldn't contact everybody using node", si)
+		} else {
+			newList = tmpList
+		}
+	}
+
+	return &CheckConnectivityReply{newList}, nil
+}
+
+func (st *Stat) testNodes(nodes []*network.ServerIdentity, to time.Duration) error {
+	r := onet.NewRoster(nodes)
+	tree := r.GenerateBinaryTree()
+	p, err := st.CreateProtocol(messaging.BroadcastName, tree)
+	if err != nil {
+		return errors.New("protocol creation failed: " + err.Error())
+	}
+	bc := p.(*messaging.Broadcast)
```

Comment:
hope this will never panic, isn't it better checking if the assertion is ok  and return a nice error if not?

Reply:
Yes, generally it's good to test if assertions are OK. But here it's just 4 lines after the `CreateProtocol`, which should definitively give back a `messaging.Broadcast` structure. Else a panic is a good idea anyway, because something is really broken and we shouldn't continue.